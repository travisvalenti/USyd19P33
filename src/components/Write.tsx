import React, { Component } from 'react'

export default class Write extends Component {
  render() {
    return (
      <div style={{
        textAlign: 'center',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: '100%'
      }}>
        <strong style={{
          display: 'inline-block',
          height: '1rem'
        }}>Haven't done this yet</strong>
      </div>
    )
  }
}
