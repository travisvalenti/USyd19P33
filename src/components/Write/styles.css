.Write {
  width: 100%;
  height: 100%;
  display: flex;
  align-content: center;
  align-items: center;
  justify-content: center;
  overflow: auto;
  box-sizing: border-box;
}

.Write .editor {
  /* height: 20rem; */
  width: 40rem;

  display: grid;
  grid-template-rows: auto auto auto;
  grid-gap: 12px;
  padding: 20px;
  box-sizing: border-box;
}

.Write .editorToolbar {
  display: flex;
  text-align: center;
  align-content: center;
  justify-content: space-evenly;
  background: white;
  box-shadow: 2px 1px 3px grey;
  border-radius: 4px;
  padding: 10px 10px 15px 10px;
}

.Write .editorToolbar .fab {
  position: relative;
  top: 0.3em;
  font-size: 1.1em;
}

.Write .editorToolbar i:hover {
  color: #57a5afee;
  cursor: pointer;
}

.Write .editorToolbar .active {
  color: #57a5af;
}

.Write .editorToolbar i {
  user-select: none;
}

.Write .editorToolbar,
.Write .textArea,
.Write .to,
.Write .subject,
.Write .from {
  padding: 10px 10px 15px 10px;
  background: white;
  box-shadow: 0 0 3px grey;
  border-radius: 4px;
  box-sizing: border-box;
}

.Write .to,
.Write .subject,
.Write .from {
  text-align: center;
  padding: 0;
  height: 2em;
}

.Write .to input,
.Write .subject input {
  height: 99%;
  width: 99%;
  background: white;
  border: none;
  margin: none;
  border-radius: inherit;
  box-sizing: border-box;
}

.Write .textArea {
  display: block;
  width: 100%;
  height: 80%;
  margin: 1% auto;
  border: none;
  min-height: 10rem;
  cursor: text;
}

.Write .mailHtml {
  background: white;
  box-shadow: 2px 1px 3px grey;
  border-radius: 4px;
  padding: 10px;
}

.Write .mailHtml p {
  display: block;
  overflow: auto;
}

.Write .actionButtons {
  display: flex;
  justify-content: space-between;
}