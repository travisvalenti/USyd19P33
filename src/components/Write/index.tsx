/** Write Component
 * This component is the sole component responsible for writing and sending mail.
 * It takes no props.
 * It manages the state using draft-js, and is also able to save drafts.
 */
import React, { Component, ChangeEvent } from 'react'
import {
  Editor,
  EditorState,
  RichUtils,
  getDefaultKeyBinding,
  ContentState,
  convertToRaw
} from 'draft-js'
import { convertToHTML } from 'draft-convert'

import './styles.css'
import Button from '../ui/Button'
import LoadingBar from '../ui/LoadingBar'

type State = {
  content: string
  markdownActive: boolean
  editorState: EditorState
  from: string
  to: string
  subject: string
  loading: boolean
}

export default class Write extends Component<{}, State> {
  state = {
    content: '',
    markdownActive: true,
    editorState: EditorState.createEmpty(),
    from: '',
    to: '',
    subject: '',
    loading: false
  }

  editorReference: any

  componentDidMount () {
    const from = this.getFromString()
    this.setState({ from })
  }

  getFromString = () => {
    const basicProfile = gapi.auth2.getAuthInstance().currentUser.get().getBasicProfile()
    return `${basicProfile.getGivenName()} ${basicProfile.getFamilyName()} <${basicProfile.getEmail()}>`
  }

  onEditorStateChange = (editorState: EditorState) => {
    this.setState({ editorState })
  }

  handleKeyCommand = (command: string, editorState: EditorState) => {
    const newState = RichUtils.handleKeyCommand(editorState, command);
    if (newState) {
      this.onEditorStateChange(newState);
      return 'handled';
    }
    return 'not-handled';
  }

  onInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    switch (event.target.name) {
      case 'to':
        this.setState({ to: event.target.value })
        break
      case 'subject':
        this.setState({ subject: event.target.value })
        break
    }
  }

  makeMime = () => {
    const currentContent = this.state.editorState.getCurrentContent()

    const blocks = convertToRaw(
      this.state.editorState.getCurrentContent()
    )
      .blocks
      .map(block => block.text)

    const content = blocks.join('\n')

    const html = convertToHTML({})(currentContent)

    let mime = `From: ${this.getFromString()}${this.state.to && '\nTo:'} ${this.state.to}${this.state.subject && '\nSubject:'} ${this.state.subject}
MIME-Version: 1.0
Content-Type: multipart/mixed; 
    boundary=gc0p4Jq0M2Yt08jU534c0p

${content}

--gc0p4Jq0M2Yt08jU534c0p
Content-Type: text/html;

${html}

--gc0p4Jq0M2Yt08jU534c0p--`

    return mime
  }

  saveDraft = () => {
    this.setState({
      loading: true
    })

    const mime = this.makeMime()

    console.log(mime)

    const response = (gapi.client as any)
      .gmail
      .users
      .drafts
      .create({
        userId: 'me',
        resource: {
          message: {
            raw: btoa(mime)
              .split('+').join('-')
              .split('/').join('_')
          }
        }
      })
    response.then(() => {
      this.setState({
        loading: false
      })
    })
      .catch((error: Error) => console.error(error))
  }

  sendEmail = () => {
    this.setState({
      loading: true
    })

    const mime = this.makeMime()

    console.log(mime)

    const response = (gapi.client as any)
      .gmail
      .users
      .messages
      .send({
        userId: 'me',
        resource: {
          raw: btoa(mime)
            .split('+').join('-')
            .split('/').join('_')
        }
      })
    response.then(() => {
        this.setState({
          loading: false
        })
    })
    .catch((error: Error) => console.error(error))
  }

  render() {
    return (
      <div className="Write">
        <div>
          <div className="editor">
            <strong>From</strong>
            <div className="from">
              { this.state.from }
            </div>
            <strong>To</strong>
            <div className="to">
              <input type="text" name="to" value={this.state.to} onChange={this.onInputChange}/>
            </div>
            <strong>Subject</strong>
            <div className="subject">
              <input type="text" name="subject" value={this.state.subject} onChange={this.onInputChange}/>
            </div>
            <strong>Message</strong>
            <div className="editorToolbar">
              <i className={ 'material-icons' } onClick={() =>
                this.onEditorStateChange(EditorState.push(this.state.editorState, ContentState.createFromText(''), 'remove-range'))
              }>delete</i>
              <i className={ 'material-icons' + (this.state.editorState.getCurrentInlineStyle().contains('ITALIC') ? ' active' : '') } onClick={() =>
                this.onEditorStateChange(RichUtils.toggleInlineStyle(this.state.editorState, 'ITALIC'))
              }>format_italic</i>
              <i className={ 'material-icons' + (this.state.editorState.getCurrentInlineStyle().contains('BOLD') ? ' active' : '') } onClick={() =>
                this.onEditorStateChange(RichUtils.toggleInlineStyle(this.state.editorState, 'BOLD'))
              }>format_bold</i>
              <i className={ 'material-icons' + (this.state.editorState.getCurrentInlineStyle().contains('STRIKETHROUGH') ? ' active' : '') } onClick={() =>
                this.onEditorStateChange(RichUtils.toggleInlineStyle(this.state.editorState, 'STRIKETHROUGH'))
              }>format_strikethrough</i>
              <i className={ 'material-icons' + (this.state.editorState.getCurrentInlineStyle().contains('UNDERLINE') ? ' active' : '') } onClick={() =>
                this.onEditorStateChange(RichUtils.toggleInlineStyle(this.state.editorState, 'UNDERLINE'))
              }>format_underline</i>
            </div>
            <div className="textArea" onClick={() =>
              this.editorReference.focus()
            }>
              <Editor
                ref={ref => this.editorReference = ref}
                editorState={this.state.editorState}
                onChange={this.onEditorStateChange}
                handleKeyCommand={this.handleKeyCommand}
                keyBindingFn={getDefaultKeyBinding}
              />
            </div>
            <div className="actionButtons">
              <Button onClick={this.saveDraft}>Save Draft</Button>
              <Button onClick={this.sendEmail}>Send</Button>
            </div>
            <div style={{ height: '1rem' }}>
              { this.state.loading &&  <LoadingBar /> }
            </div>
          </div>
        </div>
      </div>
    )
  }
}
