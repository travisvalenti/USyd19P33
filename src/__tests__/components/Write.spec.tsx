import React from 'react'
import { render, cleanup, RenderResult } from '@testing-library/react'

import Write from '../../components/Write'

describe('Write component', () => {
  afterEach(cleanup)

  test('should match snapshot', () => {
    const component = render(<Write />)
    expect(component.container).toMatchSnapshot()
  })
  
})