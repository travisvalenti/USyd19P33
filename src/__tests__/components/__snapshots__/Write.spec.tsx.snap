// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Write component should match snapshot 1`] = `
<div>
  <div
    style="text-align: center; display: flex; align-items: center; justify-content: center; min-height: 100%;"
  >
    <strong
      style="display: inline-block; height: 1rem;"
    >
      Haven't done this yet
    </strong>
  </div>
</div>
`;
